### ISSUE TITLE
Data driven new issue types

### ISSUE THREAD

**Issue by wedamija** [getsentry/sentry#43484]
We want to allow new issue types to be added without having to touch code. Currently, when we add a new type we have to touch a few places:

 - `GroupType`:  https://github.com/getsentry/sentry/blob/465f80092fd544e9479a8c99cb1a9abdaa8084a5/src/sentry/types/issues.py#L5
 - `GroupTypeToCategory`: 
 https://github.com/getsentry/sentry/blob/465f80092fd544e9479a8c99cb1a9abdaa8084a5/src/sentry/types/issues.py#L31-L42 
 - `GroupTypeToText`: https://github.com/getsentry/sentry/blob/465f80092fd544e9479a8c99cb1a9abdaa8084a5/src/sentry/types/issues.py#L44-L55
 - `GROUPHASH_IGNORE_LIMIT_MAP` https://github.com/getsentry/sentry/blob/465f80092fd544e9479a8c99cb1a9abdaa8084a5/src/sentry/event_manager.py#L167-L170

To improve this, we can move this configuration to a table. We'd have something roughly like
```
sentry_grouptype
 - id - bigint - maps to existing ids
 - slug - text - unique identifier to represnt the type
 - category - int - maps to the specific category (should category also be dynamic? might be fine for it to be hardcoded for now)
 - description - text - text descriptions from `GROUP_TYPE_TO_TEXT`
 - ignore_limit - int - values from GROUPHASH_IGNORE_LIMIT_MAP
 - enable_ingest - bool - whether we should create issues of this type (not sure if necessary, maybe it's better with rollout flags)
 - visible - bool - whether we should show these issues in the UI (again not sure if necessary, or better managed with flagr)
```

And we'd create/update these via an API? Or maybe it has to be migrations so that these issue types are created in self-hosted, ST, etc.

Kind of a pain that we'd still need to write migrations to add new types. Maybe since we still have to change code here, it makes sense just to consolidate these all into a single class?

Another option could be that the `GroupType` information is passed along with every occurrence. So then we can just upsert it with each occurrence passed. That way, we can just create these dynamically and not rely on migrations. Not sure how much I love the idea of passing this with every payload though...

**Comment by rjo100** [comment]
I think an API for custom + migrations for the ones we want could work. I worry about name collisions but could always allow updating via API.

As for ignore limit, ideally it'd have a couple options. At the very least both a limit and a duration. I'm not sure it's something we want in a database as it's currently a tool for temporarily reducing noise for new detectors.

If we want to turn it into a full fledged system that permanently introduces cutoffs for detectors, then it should go into the API. Otherwise can live in settings still but would be annoying to manually update with new detector rollouts I guess.

**Comment by mjq-sentry** [comment]
> We want to allow new issue types to be added without having to touch code.

@wedamija Would you mind expanding on why? I can certainly guess some reasons but I don't want to assume 😄 It'll help for evaluating our potential options. Thanks!

**Comment by wedamija** [comment]
> > We want to allow new issue types to be added without having to touch code.
> 
> @wedamija Would you mind expanding on why? I can certainly guess some reasons but I don't want to assume 😄 It'll help for evaluating our potential options. Thanks!

The main reasons are:
 - Internal folks who don't work on sentry would ideally not have to add boilerplate for every issue type they want to create. Although if this was the only reason, maybe it'd be fine if we just consolidated it so that you just had to register a type and provide all the config in a single place.
 - We want people external to sentry to be able to write third party detectors. It's probably not practical to require a pr for each new detector here.

**Comment by ceorourke** [comment]
**Options:**

Create a new table and pass the group type in the occurrence payload

(**Currently leaning in this direction**)

- Pros
    - No migrations or PRs needed - simplest for users
    - Can upsert data based on the payload
    - Supports self hosted, single tenant, and 3rd party contributors
- Cons
    - Need to pass same, duplicated data in payload (per issue type)
    - Risk of passing incorrect data and creating unintended group types
        - Could potentially put rate limiting on updates to prevent this

Create a new table and write migrations for new types

- Pros
    - New types are reviewed and approved (inverse of risky con on passing group type in occurrence payload)
    - Occurrence payload can be validated based on existing entries in db
    - Supports self hosted, single tenant, and 3rd party contributors
- Cons
    - Each new group type requires a migration (to support self hosted)
        - Doesn’t  initially support 3rd party contributors, but we can create API support for 3rd parties later on

Create a new table and create/update via API

- Pros
    - No duplicated data to pass in occurrence payload
- Cons
    - Doesn’t support self hosted or single tenant

Create a class

- Pros
    - Keeps the code in one place (versus currently we update 4 places)
- Cons
    - Each new type requires a PR
    - Doesn’t  initially support 3rd party contributors, but we can add support later on via API (or something else)

**Comment by mjq-sentry** [comment]
Thanks @ceorourke, and for the clarification @wedamija!

I'm trying to think through the scope of issue types and the type creation workflow to get a sense for tradeoffs.

To spitball, issue types currently exist to:
1. filter issues, via issue search
1. allow graduated feature rollout/release management of new issue types
1. apply global business rules affecting issue detection/creation (creation thresholds, throttles)
1. apply per-project/per-org business rules, exposed through the Settings tab inside Sentry _(not currently possible, but discussed)_
1. determine the appropriate issue details page UI
1. (other things I am forgetting?)

When issue creation is moved behind an API, for each of points 2-4, what is the responsibility of the caller and what is the responsibility of the issue creation API? Is the assumption that callers will only call the Create Issue API when they actually want an issue created (subject only to standard API throttling/quota limits)? Or will it be more like a "Report Issue" API that will be called more frequently, knowing that actual issue creation is still subject to type-specific configuration downstream? That might affect how much issue type data has to be stored/passed and where.

re: point 5 (and the `Each new type requires a PR` con), just calling out that in practice there's currently issue type-specific UI code for each new type, so there's at least one PR anyway. I don't know the long term goal re: customization of issue details by type, but our goal has been to make each issue as informative and actionable as possible. (Certainly there are tradeoffs and I can see the appeal of a generic page.) But if we did want to get rid of that per-type code, that might imply some other things we'd want to move into the type configuration as well.

**Comment by phacops** [comment]
@ceorourke @wedamija I'm in the camp of maximum flexibility for the user (so passing the group in the occurrence). I could see us having to define a class. Anything in between seems like a bad mix of complexity for little gain.

**Comment by mjq-sentry** [comment]
Might have buried the lede on my last one but given what we know (or think we know) now I'd lean towards a class until the third party use case is fully sorted.

Given how much back and forth actually creating and rolling out a new issue type already takes (dry-running it, tuning fingerprinting, etc), a PR to define the type does not seem like that big a lift to me compared to the complexity introduced by the other options, and unless we can get more specific about what issue type fields would have to get passed into calls to the issue API (which I'm trying to start start sorting out above) I'm concerned we'll end up with more fields than we're currently anticipating (and maybe some that doesn't seem appropriate for a caller to care about, especially a third party one) - also affecting the tradeoffs between approaches here. Just my 2 cents.

**Comment by rjo100** [comment]
Just a quick point, the issue types are used (by the performance team at least) for fingerprinting.

I agree about making things easy for 3rd parties, esp as we add more edge detection/relay level ingestions, but when we reduce the friction on ingestion we're then going to need to prompt the user to go back and customize the UI and other settings via some other endpoint. I have a very low expectation that most users will put in the necessary effort after the fact.

I think it's fine to only allow pre-registered types generally for that reason. We can even have a page in the UI that uses the endpoint for customizing UI, business rules, etc.

**Comment by ceorourke** [comment]
@mjq-sentry I was thinking the fields would be like this:

```
group_type: {
    "slug": "PERFORMANCE_SLOW_DB_QUERY",
    "category": 2,
    "description": "Slow DB Query", 
    "ignore_limit": 3,
}
```

Is it a problem to include that data every time? If it is, that's okay, that's what I'm trying to glean from people who'll actually be using this. The new table would be backfilled with all the existing `GroupType` enum data, so we wouldn't lose that for fingerprinting. The advantage here regarding 3rd parties/other teams is that they can just send along new group type information in the occurrence without having to go through another step of creating the type through the API (or whatever we choose).  I think the performance team's use case is slightly different than what we're expecting from profiling or a 3rd party for example because your code is within Sentry and you're making nice custom UI, but we're also developing a "generic" UI for new issue types that don't need (or don't want to go through the effort) customization. I really appreciate the feedback btw!!

**Comment by gggritso** [comment]
👀 @ceorourke it sounds like Performance issues might be a pretty different use-case vs. Profiling issues and potential 3rd party issues! Do you think these data-driven issue types will live side-by-side with the existing Performance issue detection pipeline, or is the goal to eventually integrate everything?

**Comment by wedamija** [comment]
> 👀 @ceorourke it sounds like Performance issues might be a pretty different use-case vs. Profiling issues and potential 3rd party issues! Do you think these data-driven issue types will live side-by-side with the existing Performance issue detection pipeline, or is the goal to eventually integrate everything?

The goal is that performance issues will migrate over to the issue platform in the next quarter or so. Ideally, at some point *all* issues will be powered by the issue platform, including error issues.

**Comment by wedamija** [comment]
> Thanks @ceorourke, and for the clarification @wedamija!
> 
> I'm trying to think through the scope of issue types and the type creation workflow to get a sense for tradeoffs.
> 
> To spitball, issue types currently exist to:
> 
> 1. filter issues, via issue search
> 2. allow graduated feature rollout/release management of new issue types
> 3. apply global business rules affecting issue detection/creation (creation thresholds, throttles)
> 4. apply per-project/per-org business rules, exposed through the Settings tab inside Sentry _(not currently possible, but discussed)_
> 5. determine the appropriate issue details page UI
> 6. (other things I am forgetting?)
> 
> When issue creation is moved behind an API, for each of points 2-4, what is the responsibility of the caller and what is the responsibility of the issue creation API? Is the assumption that callers will only call the Create Issue API when they actually want an issue created (subject only to standard API throttling/quota limits)? Or will it be more like a "Report Issue" API that will be called more frequently, knowing that actual issue creation is still subject to type-specific configuration downstream? That might affect how much issue type data has to be stored/passed and where.
> 

I think that these points are handled by our policy layer epic: https://github.com/getsentry/sentry/issues/43488.  Ideally, everyone using the issue platform will need some common functionality around rolling out new issue types. We can provide a common place for this.  There's also nothing really stopping you from using options as is to control the rollout of your detectors on your side if you still want to.

> re: point 5 (and the `Each new type requires a PR` con), just calling out that in practice there's currently issue type-specific UI code for each new type, so there's at least one PR anyway. I don't know the long term goal re: customization of issue details by type, but our goal has been to make each issue as informative and actionable as possible. (Certainly there are tradeoffs and I can see the appeal of a generic page.) But if we did want to get rid of that per-type code, that might imply some other things we'd want to move into the type configuration as well.

3rd party issue types likely won't have custom UI. You're right that for most 1st party issue types we'll end up needing to customize the ui in some way. But I'm not sure if it'll be radically different for every type. There might be a common ui for all performance issues, profiling issues, etc. And then the custom part is just the custom evidence, and descriptions/links on how to fix things.

**Comment by gggritso** [comment]
@ceorourke @wedamija gotcha, thanks! In that case, I have a few questions/comments! Sorry it's a lot, but this could be a really big change for us!

The upsert approach might be risky:
- It might require coordination between different detectors. If two detectors (3rd party, or otherwise) supply the same `slug` and/or `category` they might trample each other
- Upserting is pretty easy to do by accident (accidentally send testing events to production, accidentally rename a group type while testing, mis-type a group type string, etc.) so we might clobber our own detectors
- Also, isn't upserting pretty inefficient? We'd be doing a database write (or a read to confirm needing a write) on every occurrence!

I'm also wondering about storing detector _state_ (`enable_ingest` and `visible`) in the database. The combination of mutable state in the database and easy upserting creates some room for error:
- Sending `enable_ingest` along with occurrences can cause problems like flapping. A 3rd party integration (or even our own that runs on more than one server) can send `enable_ingest=true` from one server and `enable_ingest=false` from another because of a configuration error
- Greedy third-party integrations can choose to increase their own ignore limit!
- The detector types list is unbounded, so someone could accidentally or maliciously create a ton of detector types

I think there's a case for at least _some_ attributes to require migrations. I think there's also a case for having a separate API for managing these attributes!

Another problem Performance will run into is that having categories we can't easily control will make some important trickery much harder. For example, we test detector improvements by _subclassing_ existing detectors, and setting them to return the same group type as the superclass. This lets us run two detectors in parallel and compare their output. We deduplicate the problems, so we can easily see what new problems the subclass outputs. If the group information is stored in the database, we might not be able to do this anymore, or it'll be harder or less explicit.

I think at least _some_ detector information would be better in code, at least for the existing internal detectors.

A small note: I'm not sure if this will be the case for Profiling or 3rd party detectors, but for us it's very common to need to fetch detector configuration from the system. e.g., we have project options to control each detector's rollout so we can turn it off for users. This might come up with Profiling and 3rd party detectors, too. They may want to store and fetch detector state, which is harder for external detectors.

Maybe there's a case for being able to _fetch_ some of this info?

One more small note: Will this affect unit testing adversely? I'm guessing no, we'll just need to always mock a detector's presence in the database before testing it.

Lastly, storing the `id` and `category` in the database might make life hard on the frontend for us, because the frontend wants stable group IDs. e.g.,
- we have UI conditionals that render different UI based on group type. If you hard-code the initial set of detectors, we can keep doing this. For any new detectors, we'd have to first create the issue by sending some events, then fetch the ID, then add it to the code
- since the FE code is in TypeScript, we'd lose some typing information if the group types are not "stable"

It would be helpful if at least on the frontend the IDs were in code, since it's often hard or annoying to fetch this information dynamically.

Overall, I think this can work with Performance issues just fine, especially if we're able to keep the current detectors mostly intact. It's really exciting to potentially have async detectors that don't have to run in the middle of ingest, and call an API instead! There might be some growing pains though.

As an aside, **this sounds _a lot_ like an integration**. We can have a stable internal integration per detector declared in code, and have some infrastructure for registering 3rd party detector integrations into a database. Each integration would have a stable identity, an auth token, and store metadata and state if necessary. Then both internal and external integrations would work by sending occurrences to an API endpoint. Best of both worlds? Just thinking out loud!

**Comment by wedamija** [comment]
I kind of wish I could comment on individual points more easily, haha.

> It might require coordination between different detectors. If two detectors (3rd party, or otherwise) supply the same slug and/or category they might trample each other
I think we will need some kind of uniqueness on category/slug so that there can't be any trampling. That's something I've already been thinking about a little bit.

> Upserting is pretty easy to do by accident (accidentally send testing events to production, accidentally rename a group type while testing, mis-type a group type string, etc.) so we might clobber our own detectors
That seems like a reasonably valid concern, although isn't there also a risk of doing this via an api?

> Also, isn't upserting pretty inefficient? We'd be doing a database write (or a read to confirm needing a write) on every occurrence!
We'll always be doing a read here, at the very least from a cache. That'd be required to get the details of the group type regardless. We should rarely be performing a write though.

> Sending enable_ingest along with occurrences can cause problems like flapping. A 3rd party integration (or even our own that runs on more than one server) can send enable_ingest=true from one server and enable_ingest=false from another because of a configuration error
Yeah, I have this general concern here as well.

> Greedy third-party integrations can choose to increase their own ignore limit!
> The detector types list is unbounded, so someone could accidentally or maliciously create a ton of detector types
There will be stricter limits on what third parties can do. I don't know what these look like yet, but we'll be keeping abuse in mind.

> I think there's a case for at least some attributes to require migrations. I think there's also a case for having a separate API for managing these attributes!
For 1st party integrations this attributes can't be managed via an api, or the changes won't flow through to self-hosted/ST. The exception here would be any attributes related to releasing these in production (ie getsentry specific things).

> Another problem Performance will run into is that having categories we can't easily control will make some important trickery much harder. For example, we test detector improvements by subclassing existing detectors, and setting them to return the same group type as the superclass. This lets us run two detectors in parallel and compare their output. We deduplicate the problems, so we can easily see what new problems the subclass outputs. If the group information is stored in the database, we might not be able to do this anymore, or it'll be harder or less explicit.
That's specific to detectors. You can do whatever you want in your detector code, issue platform doesn't care at all. If you want to have an enum of all the types you support and use that in some way to produce different detections that's fine and doesn't impact us at all.

> A small note: I'm not sure if this will be the case for Profiling or 3rd party detectors, but for us it's very common to need to fetch detector configuration from the system. e.g., we have project options to control each detector's rollout so we can turn it off for users. This might come up with Profiling and 3rd party detectors, too. They may want to store and fetch detector state, which is harder for external detectors.
I think that https://github.com/getsentry/sentry/issues/43488 helps solve this, but let me know if I'm missing something.

> Lastly, storing the id and category in the database might make life hard on the frontend for us, because the frontend wants stable group IDs. e.g.,
> we have UI conditionals that render different UI based on group type. If you hard-code the initial set of detectors, we can keep doing this. For any new detectors, we'd have to first create the issue by sending some events, then fetch the ID, then add it to the code
> since the FE code is in TypeScript, we'd lose some typing information if the group types are not "stable"

I'd say you should rely on the slugs (or category+slug) instead of the ids here.

> As an aside, this sounds a lot like an integration. We can have a stable internal integration per detector declared in code, and have some infrastructure for registering 3rd party detector integrations into a database. Each integration would have a stable identity, an auth token, and store metadata and state if necessary. Then both internal and external integrations would work by sending occurrences to an API endpoint. Best of both worlds? Just thinking out loud!

Something like this probably makes sense for at least 3rd party detectors.

So just want to say that we're not married to the "send info with occurrences" idea. I don't love it for a lot of the reasons you mentioned, but wanted to avoid folks having to make prs that create migrations to make new group types whenever they want to add new things. 

Another potential option is a hybrid: 1st party integrations just register a dataclass with the relevant information, and third parties use the database. Then we have an abstraction around this so that there's no distinction between the two storage methods in terms of how we reason about them.

**Comment by mjq-sentry** [comment]
> Another potential option is a hybrid: 1st party integrations just register a dataclass with the relevant information, and third parties use the database. Then we have an abstraction around this so that there's no distinction between the two storage methods in terms of how we reason about them.

FWIW I like that approach!

I agree that for third parties, a database-backed approach makes sense. Those are dynamic and shouldn't need our involvement or blessing like a PR would. As for how that data is managed, I've been swayed by @gggritso's points that compared to piggybacking off of the issue API, a separate registration step (in the UI or via an API) for the issue type/detector seems less error-prone, while not adding much burden on third-party detector implementers. I'm assuming that those implementers have to do similar setup for a webhook or whatever other mechanism they're going to use to process events for issues anyway. (Or maybe not, if the idea is for them to use their own Relay instance?). That said, type info as params to the issue API still sounds workable and @wedamija dealt with some of my concerns above too.

My concerns with a database-backed approach are more for first party issues. We already have access to the event firehose, and we don't think about registering or managing types the same way as a third party would - for example, outside the rollout period we typically want to control them globally, not on a per-organization basis. We could manage them directly in the database through migrations, but that makes the data harder to see and verify - with our current combination of code and flagger it's easy to verify the state. It's tougher from the database. If migrations are all PR'd and kept in the Sentry repo then we have some auditability there, but then we've got the same "PR required for every new detector" requirement as with storing the config in code.

That leaves consistency as the main benefit going for a DB-for-first-party solution, but to me the use cases and concerns are subtly different enough that different implementations seems okay to me (while not so divergent that we couldn't put a common interface overtop). We also know we'll be building a lot of detectors and understand our own use case well, while the third party case is not as well understood yet.

Anyway, it's up to you to build this, not me! Just my 2 cents. As far as I can tell, all the performance team _really_ needs is stable issue type identifiers and the means to control both rollout and issue thresholds, either upstream or downstream of the issue API. @wedamija covered these points a lot in his last few replies and I don't have any concerns there.

Thanks!

**Comment by gggritso** [comment]
@wedamija I think that's all fair enough, and makes sense! FWIW I also like the hybrid approach, since it both keeps things simpler for now _and_ punts the 3rd-party question until later 😂

**Comment by ceorourke** [comment]
We discussed this in a meeting and decided on a dataclass approach for the short term, and we'll figure out a better solution for 3rd parties later on as we understand that requirement better.
